Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,386 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP1
+import NIOSOCKS
+import NIOSSL
+import NIOTLS
+#if canImport(Network)
+    import NIOTransportServices
+#endif
+
+extension HTTPConnectionPool {
+    enum NegotiatedProtocol {
+        case http1_1(Channel)
+        case http2_0(Channel)
+    }
+
+    final class ConnectionFactory {
+        struct Timeouts {
+            /// the `TimeAmount` waited before the TCP connection creation is failed with a timeout error
+            var connect: TimeAmount = .seconds(10)
+
+            /// the `TimeAmount` waited before the SOCKS proxy connection creation is failed with a timeout error
+            var socksProxyHandshake: TimeAmount = .seconds(10)
+
+            /// the `TimeAmount` waited before the HTTP proxy connection creation is failed with a timeout error
+            var httpProxyHandshake: TimeAmount = .seconds(10)
+
+            /// the `TimeAmount` waited before we the TLS handshake is failed with a timeout error
+            var tlsHandshake: TimeAmount = .seconds(10)
+        }
+
+        let key: ConnectionPool.Key
+        let clientConfiguration: HTTPClient.Configuration
+        let tlsConfiguration: TLSConfiguration
+        let sslContextCache: SSLContextCache
+        let timeouts: Timeouts
+
+        init(key: ConnectionPool.Key,
+             tlsConfiguration: TLSConfiguration?,
+             clientConfiguration: HTTPClient.Configuration,
+             sslContextCache: SSLContextCache,
+             timeouts: Timeouts = .init()) {
+            self.key = key
+            self.clientConfiguration = clientConfiguration
+            self.sslContextCache = sslContextCache
+            self.tlsConfiguration = tlsConfiguration ?? clientConfiguration.tlsConfiguration ?? .forClient()
```

Comment:
So if we have both a `tlsconfigurataion` and `clientConfiguration.tlsConfiguration`, we drop the client configuration on the floor? Why? Might be useful to have a comment

Reply:
This prefers _either_ an explicit `tlsConfiguration`, or if that's not available, one from `clientConfiguration`, or if there isn't one there either it constructs a plain one.